File: frontend/src/lib/stockPlacements.ts
export type StockPlacement = {
  location: string;
  quantity: number;
  quality?: string;
  zone?: string;
  aisle?: string;
  shelf?: string;
  box?: string;
  storage_code?: string;
  notes?: string;
};

export type StockPlacementFormRow = {
  location: string;
  quantity: string;
  quality: string;
  zone: string;
  aisle: string;
  shelf: string;
  box: string;
  storage_code: string;
  notes: string;
};

function trimOpt(value: unknown): string | undefined {
  if (typeof value !== 'string') return undefined;
  const t = value.trim();
  return t.length > 0 ? t : undefined;
}

export function formatPlacementLabel(placement: StockPlacement): string {
  const parts: string[] = [placement.location];
  if (placement.quality) parts.push(placement.quality);
  if (placement.zone) parts.push(placement.zone);
  if (placement.aisle) parts.push(`Aisle ${placement.aisle}`);
  if (placement.shelf) parts.push(`Shelf ${placement.shelf}`);
  const bin = placement.box || placement.storage_code;
  if (bin) {
    parts.push(bin.startsWith('Box') || bin.startsWith('Cell') || bin.startsWith('Tote') ? bin : `Box ${bin}`);
  }
  return parts.join(' · ');
}

export function placementSearchBlob(placement: StockPlacement): string {
  return [
    placement.location,
    placement.quality,
    placement.zone,
    placement.aisle,
    placement.shelf,
    placement.box,
    placement.storage_code,
    placement.notes,
    formatPlacementLabel(placement),
  ]
    .filter(Boolean)
    .join(' ')
    .toLowerCase();
}

export function parsePlacementsFromItem(item: {
  locations?: StockPlacement[];
  location?: string;
  quantity?: number;
  quality?: string;
}): StockPlacement[] {
  if (item.locations && item.locations.length > 0) {
    return item.locations.map((raw) => ({
      location: raw.location || item.location || 'Store',
      quantity: Number(raw.quantity) || 0,
      quality: raw.quality || item.quality || 'New',
      zone: trimOpt(raw.zone),
      aisle: trimOpt(raw.aisle),
      shelf: trimOpt(raw.shelf),
      box: trimOpt(raw.box),
      storage_code: trimOpt(raw.storage_code),
      notes: trimOpt(raw.notes),
    }));
  }
  return [{ location: item.location || 'Store', quantity: Number(item.quantity) || 0, quality: item.quality || 'New' }];
}

export function placementFormFromApi(raw: StockPlacement): StockPlacementFormRow {
  return {
    location: raw.location,
    quantity: String(raw.quantity),
    quality: raw.quality || 'New',
    zone: raw.zone || '',
    aisle: raw.aisle || '',
    shelf: raw.shelf || '',
    box: raw.box || '',
    storage_code: raw.storage_code || '',
    notes: raw.notes || '',
  };
}

export function emptyPlacementFormRow(defaultLocation: string): StockPlacementFormRow {
  return {
    location: defaultLocation,
    quantity: '0',
    quality: 'New',
    zone: '',
    aisle: '',
    shelf: '',
    box: '',
    storage_code: '',
    notes: '',
  };
}

export function placementFormToApi(row: StockPlacementFormRow): StockPlacement {
  const placement: StockPlacement = {
    location: row.location,
    quantity: parseInt(row.quantity, 10) || 0,
    quality: row.quality.trim() || 'New',
  };
  const zone = row.zone.trim();
  const aisle = row.aisle.trim();
  const shelf = row.shelf.trim();
  const box = row.box.trim();
  const storage_code = row.storage_code.trim();
  const notes = row.notes.trim();
  if (zone) placement.zone = zone;
  if (aisle) placement.aisle = aisle;
  if (shelf) placement.shelf = shelf;
  if (box) placement.box = box;
  if (storage_code) placement.storage_code = storage_code;
  if (notes) placement.notes = notes;
  return placement;
}

/** Level mode tracks Low/Med/High — placements store location only (qty 0). */
export function placementRowsToApi(
  rows: StockPlacementFormRow[],
  mode: QuantityMode,
): StockPlacement[] {
  const parsed = rows.map(placementFormToApi);
  if (mode === 'level') {
    return parsed.map((p) => ({ ...p, quantity: 0 }));
  }
  return parsed;
}

export function pickDefaultPlacementIndex(placements: StockPlacement[]): number {
  if (placements.length === 0) return 0;
  let bestIdx = 0;
  let bestQty = placements[0]?.quantity ?? 0;
  for (let i = 1; i < placements.length; i++) {
    const q = placements[i]?.quantity ?? 0;
    if (q > bestQty) {
      bestQty = q;
      bestIdx = i;
    }
  }
  return bestIdx;
}

export function validatePlacementsRequireBin(
  placements: StockPlacement[],
  requireBinForLocations: string[],
): string | null {
  const requireSet = new Set(requireBinForLocations.map((l) => l.toLowerCase()));
  for (const p of placements) {
    if (p.quantity <= 0) continue;
    if (!requireSet.has(p.location.toLowerCase())) continue;
    if (!p.box && !p.storage_code) {
      return `Box or storage code is required for ${p.location} placements`;
    }
  }
  return null;
}

export type QuantityMode = 'count' | 'level';
export type QuantityLevel = 'low' | 'medium' | 'high';

export const QUANTITY_LEVELS: { value: QuantityLevel; label: string }[] = [
  { value: 'low', label: 'Low' },
  { value: 'medium', label: 'Medium' },
  { value: 'high', label: 'High' },
];

export function quantityLevelLabel(level: string | null | undefined): string {
  if (!level) return '';
  const found = QUANTITY_LEVELS.find((l) => l.value === level);
  return found ? found.label : level;
}

export function quantityLevelColor(level: string | null | undefined): string {
  switch (level) {
    case 'low': return 'text-rose-600 bg-rose-50 border-rose-200';
    case 'medium': return 'text-amber-600 bg-amber-50 border-amber-200';
    case 'high': return 'text-emerald-600 bg-emerald-50 border-emerald-200';
    default: return 'text-slate-600 bg-slate-50 border-slate-200';
  }
}

File: mobile_app/lib/core/stock_placements.dart
String formatPlacementLabel(Map<String, dynamic> placement) {
  final parts = <String>[placement['location'] as String? ?? 'Store'];
  final quality = (placement['quality'] as String?)?.trim();
  if (quality != null && quality.isNotEmpty) parts.add(quality);
  final zone = (placement['zone'] as String?)?.trim();
  final aisle = (placement['aisle'] as String?)?.trim();
  final shelf = (placement['shelf'] as String?)?.trim();
  final box = ((placement['box'] as String?)?.trim().isNotEmpty == true
          ? placement['box']
          : placement['storage_code'])
      as String?;
  if (zone != null && zone.isNotEmpty) parts.add(zone);
  if (aisle != null && aisle.isNotEmpty) parts.add('Aisle $aisle');
  if (shelf != null && shelf.isNotEmpty) parts.add('Shelf $shelf');
  if (box != null && box.isNotEmpty) {
    parts.add(box.startsWith('Box') || box.startsWith('Cell') || box.startsWith('Tote') ? box : 'Box $box');
  }
  return parts.join(' · ');
}

List<Map<String, dynamic>> parsePlacementsFromItem(Map<String, dynamic> item) {
  final raw = item['locations'];
  if (raw is List && raw.isNotEmpty) {
    return raw.map((entry) {
      if (entry is! Map) return <String, dynamic>{'location': 'Store', 'quantity': 0, 'quality': 'Used - Good'};
      return {
        'location': entry['location'] as String? ?? item['location'] as String? ?? 'Store',
        'quantity': (entry['quantity'] as num?)?.toInt() ?? 0,
        'quality': entry['quality'] as String? ?? item['quality'] as String? ?? 'Used - Good',
        if (entry['zone'] != null) 'zone': '${entry['zone']}',
        if (entry['aisle'] != null) 'aisle': '${entry['aisle']}',
        if (entry['shelf'] != null) 'shelf': '${entry['shelf']}',
        if (entry['box'] != null) 'box': '${entry['box']}',
        if (entry['storage_code'] != null) 'storage_code': '${entry['storage_code']}',
        if (entry['notes'] != null) 'notes': '${entry['notes']}',
      };
    }).toList();
  }
  return [
    {
      'location': item['location'] as String? ?? 'Store',
      'quantity': (item['quantity'] as num?)?.toInt() ?? 0,
      'quality': item['quality'] as String? ?? 'Used - Good',
    },
  ];
}

String placementSearchBlob(Map<String, dynamic> placement) {
  return [
    placement['location'],
    placement['quality'],
    placement['zone'],
    placement['aisle'],
    placement['shelf'],
    placement['box'],
    placement['storage_code'],
    placement['notes'],
    formatPlacementLabel(placement),
  ].where((v) => v != null && '$v'.trim().isNotEmpty).join(' ').toLowerCase();
}

Map<String, dynamic> emptyPlacementRow(String defaultLocation) {
  return {
    'location': defaultLocation,
    'quantity': 0,
    'quality': 'Used - Good',
    'zone': '',
    'aisle': '',
    'shelf': '',
    'box': '',
    'storage_code': '',
    'notes': '',
  };
}

Map<String, dynamic> placementRowToApi(Map<String, dynamic> row) {
  final placement = <String, dynamic>{
    'location': row['location'] as String? ?? 'Store',
    'quantity': (row['quantity'] as num?)?.toInt() ?? int.tryParse('${row['quantity']}') ?? 0,
    'quality': row['quality'] as String? ?? 'Used - Good',
  };
  for (final key in ['zone', 'aisle', 'shelf', 'box', 'storage_code', 'notes']) {
    final val = (row[key] as String?)?.trim() ?? '';
    if (val.isNotEmpty) placement[key] = val;
  }
  return placement;
}

List<Map<String, dynamic>> placementRowsToApi(
  List<Map<String, dynamic>> rows,
  String quantityMode,
) {
  final parsed = rows.map(placementRowToApi).toList();
  if (quantityMode == 'level') {
    return parsed.map((p) => {...p, 'quantity': 0}).toList();
  }
  return parsed;
}

String? validatePlacementsRequireBin(List<Map<String, dynamic>> placements, List<String> requireSites) {
  final require = requireSites.map((s) => s.toLowerCase()).toSet();
  for (final p in placements) {
    final qty = (p['quantity'] as num?)?.toInt() ?? 0;
    if (qty <= 0) continue;
    final site = (p['location'] as String? ?? '').toLowerCase();
    if (!require.contains(site)) continue;
    final box = (p['box'] as String?)?.trim() ?? '';
    final code = (p['storage_code'] as String?)?.trim() ?? '';
    if (box.isEmpty && code.isEmpty) {
      return 'Box or storage code required for ${p['location']}';
    }
  }
  return null;
}

int pickDefaultPlacementIndex(List<Map<String, dynamic>> placements) {
  if (placements.isEmpty) return 0;
  var bestIdx = 0;
  var bestQty = (placements[0]['quantity'] as num?)?.toInt() ?? 0;
  for (var i = 1; i < placements.length; i++) {
    final q = (placements[i]['quantity'] as num?)?.toInt() ?? 0;
    if (q > bestQty) {
      bestQty = q;
      bestIdx = i;
    }
  }
  return bestIdx;
}

const jobPartStatuses = <String>[
  'requested',
  'on_order',
  'available',
  'picked_up',
  'installed',
  'cancelled',
  'returned',
];

String jobPartStatusLabel(String status) {
  switch (status) {
    case 'on_order':
      return 'On order';
    case 'picked_up':
      return 'Picked up';
    default:
      return status.replaceAll('_', ' ').split(' ').map((w) {
        if (w.isEmpty) return w;
        return '${w[0].toUpperCase()}${w.substring(1)}';
      }).join(' ');
  }
}

const quantityLevels = ['low', 'medium', 'high'];

String quantityLevelLabel(String? level) {
  switch (level) {
    case 'low':
      return 'Low';
    case 'medium':
      return 'Medium';
    case 'high':
      return 'High';
    default:
      return level ?? '';
  }
}
